### PR TITLE
fix(MultiSelect): handle case of MultiSelect with initialSelectedItems

### LINKED ISSUE #707: [Regression] MultiSelect does not un-select properly
Seems like we might have had a regression recently with the way MultiSelect handles selection. Including a gif to reproduce below:

kapture 2018-03-13 at 9 08 48

### PR DESCRIPTION
**Pull request by rezak-otmani**
Closes #707  and #708

`componentWillReceiveProps` re-insert the removed items systematically

#### Changelog

**New**

**Changed**
**Removed**
`componentWillReceiveProps`

### CODE REVIEW

**Comment by joshblack** [comment]
Hi @rezak-otmani! 👋 Thanks for taking the time to make this PR! One last note before reviewing, could you add test cases for the two issues to help protect against these regressions in the future? Thanks!

**Comment by asudoh** [comment]
Hi @rezak-otmani thank you for trying to address the problem, one quick question - Did you intend to kill `prop` -> `state` sync for the component?

**Comment by rezak-otmani** [comment]
Hi @asudoh not sure understand what you mean, but I think that I do not intend to do it

**Comment by asudoh** [comment]
@rezak-otmani OK - Then I believe `cWRP` code you removed should come back. Do you want to work on this? Thanks!

**Comment by rezak-otmani** [comment]
Hi @asudoh I modifed the PR, just changing the condition `(itemsToAdd.length> 0)` which is still true as long as there are the initialSelectedItems 
this condition is modifed in a way that `cWRP` does not re-insert the elements de- inserted

**Comment by asudoh** [comment]
@joshblack Would you agree with the latest fix? Looks simple by scratching the surface, but need expertise from one who deeply understands the codebase. Thanks!

File: src/internal/Selection.js [hunk 1]
```diff
@@ -23,7 +23,7 @@ export default class Selection extends React.Component {
       return selectedItems.indexOf(item) === -1;
     });
 
-    if (itemsToAdd.length > 0) {
+    if (selectedItems.indexOf(itemsToAdd) !== -1) {
```
Comment (joshblack) [inline src/internal/Selection.js]: In this block, would `itemsToAdd` be an array? Seems like this branch would be unreachable as a result. Do we need it then?

Comment (rezak-otmani) [reply src/internal/Selection.js]: yes, `itemsToAdd` is an array that contains `initialSelectedItems` de-inserted, 
so the contents of `itemsToAddd` should not be re-inserted,
I think adding items here must be removed,  but what does do `cWRP` then

Comment (joshblack) [reply src/internal/Selection.js]: Potentially nothing. I think the goal here originally was to select things that were added to `initialSelectedItems` but weren't in `selectedItems`. If that branch isn't one that makes sense, can totally remove this 😄

**Comment by asudoh** [comment]
@rezak-otmani Seems that `cWRP` is removed again? Would you want to take another look to see what happened with the merging? Thanks!

**Comment by rezak-otmani** [comment]
@asudoh `cWRP` has no impact on other components except the MultiSelect, and it does not have any sense even for this component, so I do not see the usefulness of keeping it

**Comment by asudoh** [comment]
OK thanks @rezak-otmani for clarifying! It's fine as long as @joshblack is good with it, but another thing that makes me wonder is removal of `cWRP` is the only change here...?

**Comment by rezak-otmani** [comment]
yes,  cWRP is the only cause of the issue

**Comment by asudoh** [comment]
Thanks @rezak-otmani for more clarification! My last question is - Will we support changing list of selected items via `props`?

**Comment by rezak-otmani** [comment]
I think it's already done via  state

**Comment by asudoh** [comment]
Just double-checking - Do you think the new code covers the following scenario `props.initialSelectedItems` being `[{ id: 'id-foo', label: 'label-foo', value: 'value-foo' }, { id: 'id-bar', label: 'label-bar', value: 'value-bar' }]` initially and then changed to `[{ id: 'id-foo', label: 'label-foo', value: 'value-foo' }, { id: 'id-baz', label: 'label-baz', value: 'value-baz' }, { id: 'id-baz', label: 'label-bar', value: 'value-bar' }]`? You said it's done via state, but wanted to see how. Thanks!

**Comment by joshblack** [comment]
@asudoh if I understand correctly, I think the intent behind the `cWRP` code was to update selected items with what was in `initialSelectedItems` if it was updated. I think the naming of that prop suggests that it should only apply to first use (e.g. the initial stage) which makes this patch potentially valid in addition to fixing the current issue people are facing with MultiSelect. Definitely was an error on my side with the initial component logic here, I think 😅

**Comment by asudoh** [comment]
OK @joshblack thank you for clarifying!!

**Pull request merged**